algorithms/a3c/model.py

```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+        super(A3C_LSTM_GA, self).__init__()
+
+        # Image Processing
+        self.conv1 = nn.Conv2d(num_input_channels, 128, kernel_size=8, stride=4)
+        self.conv2 = nn.Conv2d(128, 64, kernel_size=4, stride=2)
+        self.conv3 = nn.Conv2d(64, 64, kernel_size=4, stride=2)
+
+        self.output_width, self.output_height, \
+            self.num_output_filters = calculate_input_width_height_for_A3C_LSTM_GA(resolution)
+        self.lstm_cell_size = self.output_width * self.output_height * self.num_output_filters
+
+        # Instruction Processing
+        self.gru_hidden_size = 256
+        self.input_size = vocab_size
+        self.embedding = nn.Embedding(self.input_size, 32)
+        self.gru = nn.GRU(32, self.gru_hidden_size)
+
+        # Gated-Attention layers
+        self.attn_linear = nn.Linear(self.gru_hidden_size, 64)
+
+        # Time embedding layer, helps in stabilizing value prediction
+        self.time_emb_dim = 32
+        self.time_emb_layer = nn.Embedding(
+                episode_length,
+                self.time_emb_dim)
+
+        # A3C-LSTM layers
+        self.linear = nn.Linear(self.lstm_cell_size, 256)
+        self.lstm = nn.LSTMCell(256, 256)
+        self.critic_linear = nn.Linear(256 + self.time_emb_dim, 1)
+        self.actor_linear = nn.Linear(256 + self.time_emb_dim, num_outputs)
+
+        # Initializing weights
+        self.apply(weights_init)
+        self.actor_linear.weight.data = normalized_columns_initializer(
+            self.actor_linear.weight.data, 0.01)
+        self.actor_linear.bias.data.fill_(0)
+        self.critic_linear.weight.data = normalized_columns_initializer(
+            self.critic_linear.weight.data, 1.0)
+        self.critic_linear.bias.data.fill_(0)
+
+        self.lstm.bias_ih.data.fill_(0)
+        self.lstm.bias_hh.data.fill_(0)
+        self.train()
+
+    def forward(self, inputs):
+        x, input_inst, (tx, hx, cx) = inputs
+
+        # Get the image representation
+        x = F.relu(self.conv1(x))
+        x = F.relu(self.conv2(x))
+        x_image_rep = F.relu(self.conv3(x))
+
+        # Get the instruction representation
+        encoder_hidden = torch.zeros(1, 1, self.gru_hidden_size)
+        for i in range(input_inst.data.size(1)):
+            word_embedding = self.embedding(input_inst[0, i]).view(1, 1, -1)  # 1x1x32
+            _, encoder_hidden = self.gru(word_embedding, encoder_hidden)
+        x_instr_rep = encoder_hidden.view(encoder_hidden.size(1), -1)
+
+        # Get the attention vector from the instruction representation
+        x_attention = F.sigmoid(self.attn_linear(x_instr_rep))
+
+        # Gated-Attention
+        x_attention = x_attention.unsqueeze(2).unsqueeze(3)
+        x_attention = x_attention.expand(1, self.num_output_filters, self.output_height,
+                                         self.output_width)
+        assert x_image_rep.size() == x_attention.size()
```

Comment:
No asserts. Check and raise if it doesn't work

Reply:
hmmmm, i really like the conciseness of this assert and i can add a comment?
check and raise is lame to have in a forward function? 2 lines and then you have ugly raise ValueError's in your beauitfully clear forward function